src/types/string.rs

```diff
@@ -238,6 +221,138 @@ impl PyString {
+
+    /// Converts the `PyString` into a Rust string, avoiding copying when possible.
+    ///
+    /// Returns a `UnicodeEncodeError` if the input is not valid unicode
+    /// (containing unpaired surrogates).
+    fn to_cow(&self) -> PyResult<Cow<'_, str>>;
+
+    /// Converts the `PyString` into a Rust string.
+    ///
+    /// Unpaired surrogates invalid UTF-8 sequences are
+    /// replaced with `U+FFFD REPLACEMENT CHARACTER`.
+    fn to_string_lossy(&self) -> Cow<'_, str>;
+
+    /// Obtains the raw data backing the Python string.
+    ///
+    /// If the Python string object was created through legacy APIs, its internal storage format
+    /// will be canonicalized before data is returned.
+    ///
+    /// # Safety
+    ///
+    /// This function implementation relies on manually decoding a C bitfield. In practice, this
+    /// works well on common little-endian architectures such as x86_64, where the bitfield has a
+    /// common representation (even if it is not part of the C spec). The PyO3 CI tests this API on
+    /// x86_64 platforms.
+    ///
+    /// By using this API, you accept responsibility for testing that PyStringData behaves as
+    /// expected on the targets where you plan to distribute your software.
+    #[cfg(not(Py_LIMITED_API))]
+    unsafe fn data(&self) -> PyResult<PyStringData<'_>>;
+}
+
+impl<'py> PyStringMethods<'py> for Py2<'py, PyString> {
+    #[cfg(any(Py_3_10, not(Py_LIMITED_API)))]
+    fn to_str(&self) -> PyResult<&str> {
+        Py2Borrowed::from(self).to_str()
+    }
+
+    fn to_cow(&self) -> PyResult<Cow<'_, str>> {
+        Py2Borrowed::from(self).to_cow()
+    }
+
+    fn to_string_lossy(&self) -> Cow<'_, str> {
+        Py2Borrowed::from(self).to_string_lossy()
+    }
+
+    #[cfg(not(Py_LIMITED_API))]
+    unsafe fn data(&self) -> PyResult<PyStringData<'_>> {
+        Py2Borrowed::from(self).data()
+    }
+}
+
+impl<'a> Py2Borrowed<'a, '_, PyString> {
+    #[cfg(any(Py_3_10, not(Py_LIMITED_API)))]
+    #[allow(clippy::wrong_self_convention)]
+    fn to_str(self) -> PyResult<&'a str> {
+        // PyUnicode_AsUTF8AndSize only available on limited API starting with 3.10.
+        let mut size: ffi::Py_ssize_t = 0;
+        let data: *const u8 =
+            unsafe { ffi::PyUnicode_AsUTF8AndSize(self.as_ptr(), &mut size).cast() };
+        if data.is_null() {
+            Err(crate::PyErr::fetch(self.py()))
+        } else {
+            Ok(unsafe {
+                std::str::from_utf8_unchecked(std::slice::from_raw_parts(data, size as usize))
+            })
+        }
+    }
+
+    #[allow(clippy::wrong_self_convention)]
+    fn to_cow(self) -> PyResult<Cow<'a, str>> {
+        // TODO: this method can probably be deprecated once Python 3.9 support is dropped,
+        // because all versions then support the more efficient `to_str`.
+
+        let ptr = self.as_ptr();
+        let py = self.py();
+
+        #[cfg(any(Py_3_10, not(Py_LIMITED_API)))]
+        if let Ok(s) = self.to_str() {
```

Comment:
Couldn't this error be forwarded (in contrast to `to_string_lossy`) because `PyUnicode_AsUTF8String` cannot succeed if the UTF-8 check already failed?

Reply:
Yes, I think you are right, and that would also improve coverage.